Add SyncAdapter support

zsiegel opened the pull request:
Pull request for issue discussed in #52

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
Adding support in the library for registering content observers based on the Model class

Commit:
adding syncing capability to the sample application

emilsjolander commented on library/src/main/java/se/emilsjolander/sprinkles/SprinklesContentObserver.java:
```diff
@@ -0,0 +1,54 @@
+     * @param authority the authority of the ContentProvider
+     * @return a ContentObserver
+     */
+    public static ContentObserver observer(final Account account, final String authority) {
```
I don't really see a reason for this as aposed to SprinklesContentObserver being a subclass of ContentObserver.

zsiegel replied:
I actually had this as a separate object like you suggested and then refactored into this static method. I will go ahead and update this.

emilsjolander commented on library/src/main/java/se/emilsjolander/sprinkles/Sprinkles.java:
```diff
@@ -8,21 +12,14 @@
+
+    Context mContext;
+    List<Migration> mMigrations = new ArrayList<Migration>();
+    Map<Class, ContentObserver> observers = new ConcurrentHashMap<Class, ContentObserver>();
```
Why should we keep references to the observers?

emilsjolander commented on library/src/test/java/se/emilsjolander/sprinkles/SprinklesTest.java:
```diff
@@ -63,6 +72,25 @@ public void addMigration() {
 
     @Test
+    public void addObserver() {
+        Sprinkles s = Sprinkles.init(Robolectric.application);
```
this only tests that the observer is added to our internal list not actually is registered as to the content resolver

emilsjolander commented:
Something about this should be added in the README as well. Also i am unsure about including this in the sample application just because it increases the complexity of the sample by quite a bit without showing of a lot of sprinkles specific stuff.

Otherwise i think it is a great solution to the issue! :+1:

zsiegel commented on library/src/main/java/se/emilsjolander/sprinkles/Sprinkles.java:
```diff
@@ -104,4 +101,17 @@ TypeSerializer getTypeSerializer(Class<?> type) {
+        observers.put(clazz, observer);
+    }
+
+    public void removeContentObserver(Class<? extends Model> clazz) {
```
We need a reference to the observers in order to remove them later on if needed

zsiegel replied:
Actually on second thought let me refactor this. 

It would make more sense to just return the SprinklesContentObserver and let the user handle them however they wish.

emilsjolander replied:
Ok, thats true, i missed that.

Commit:
refactoring SyncAdapter support
- the Sprinkles instance now returns ContentObservers for the user to manage
- the README has been updated to include a sample of how to use the functionality
- tests have been updated

zsiegel commented:
Ok so I have refactored things a bit.

You questioned why we needed to keep references to the ContentObservers. After re-thinking this a bit we don't and shouldn't be. The Sprinkles instance the user creates will now simple return those ContentObserver objects and they will manage them.

One side effect of this was I had to make the Utils class public and then scoped the `getNotificationURI()` method as public as well. We could just expose this on the Model class instead as a static method. either way would be fine with me.

I have also updated the README.md file with an example of how to use the observers and have cleaned up the sample code.

jdreesen commented on README.md:
```diff
@@ -246,3 +246,19 @@ Through an instance of `Sprinkles` you can register your own `TypeSerializer` in
+protected void onResume() {
+    super.onResume();
+    this.contentObserver = sprinkles.getContentObserver(getSyncUser(), AuthConstants.CONTENT_AUTHORITY);
+    sprinkles.getContentResolver().registerContentObserver(Utils.getNotificationUri(Note.class), true, contentObserver);
```
Why not add a proxy method for convenience?

``` java
public void registerContentObserver(Model model, boolean notifyForDescendents, ContentObserver observer) {
    sInstance.mContext.getContentResolver().registerContentObserver(Utils.getNotificationUri(model), true, observer);
}
```

This way there's no need to change the scope of the `Utils` class.

zsiegel replied:
I thought about this long and hard... The issue is in order not to leak anything the user of the library needs to unregister the observers within their lifecycle methods.

We could add that proxy method but change the signature to return the observer that was created and registered if this is desired.

jdreesen replied:
My intention was to add this method in addition to `public static ContentObserver getContentObserver(Account account, String authority)` (and maybe add a simple proxy for `unregisterContentObserver()` as well) so the user could call it like this:

``` java
this.contentObserver = sprinkles.getContentObserver(getSyncUser(), AuthConstants.CONTENT_AUTHORITY);
sprinkles.registerContentObserver(Note.class, true, this.contentObserver);
```

and could unregister like this:

``` java
sprinkles.unregisterContentObserver(this.contentObserver);
```

zsiegel replied:
This is a pretty good idea. ill wait for @emilsjolander to weigh in but this would clean things up even further and not require any changes to the Utils class visibility.

jdreesen replied:
And maybe the third parameter of `registerContentObserver()` should be a `SprinklesContentObserver` instead of just `ContentObserver`

emilsjolander replied:
I like where this is going but would like to change one thing. The current solution would force the user to get a `Sprinkles` instance in their application code which is not how the rest of the API works. How about doing something like this instead?

``` java
ContentObserver myPlainObserver = ...;
this.observer = new SprinklesContentObserver(myPlainObserver);

@Override
public void onResume() {
    this.observer.observe(Note.class, true); // true/false for notify descendants
}

@Override
public void onPause() {
    this.observer.unobserve(); // unobserve is a weird name but you get my point
}
```

Commit:
refactoring the observation
- users now create their own SprinklesObserver and then register that object for observation
- put Utils back to the way it was originally

zsiegel commented:
I have refactored how the user observes their models and updated the README accordingly.

The user will now create their own `SprinklesContentObserver` as @emilsjolander suggested and then they can call `register()` and `unregister()` as they see fit.

jdreesen commented on README.md:
````diff
@@ -246,3 +246,20 @@ Through an instance of `Sprinkles` you can register your own `TypeSerializer` in
+###SyncAdapters
+Sprinkles supports notifying SyncAdapters that the content has changed. By registering your models for notifications you can ensure your SyncAdapter will be notified.
+```java
+ContentObserver observer;
````
Nice :) but I think this should be a `SprinklesContentObserver` because there are no `register()` and `unregister()` methods in `ContentObserver`.

zsiegel replied:
Yes good catch! Ill update that on the next push.

jdreesen commented on README.md:
```diff
@@ -246,3 +246,20 @@ Through an instance of `Sprinkles` you can register your own `TypeSerializer` in
+this.observer = new SprinklesContentObserver(account, authority);
+
+@Override
+public void onResume() {
```
You should call `super.onResume()` here, no?

jdreesen commented on README.md:
```diff
@@ -246,3 +246,20 @@ Through an instance of `Sprinkles` you can register your own `TypeSerializer` in
+}
+
+@Override
+public void onPause() {
```
And here too.

Commit:
edits to the README code on SyncAdapter support

zsiegel commented:
Updated the README. I left the `super.***()`  calls for brevity but I can see how people might just copy and paste the methods and get screwed up.

emilsjolander commented:
i like it :+1: Make sure to update the tests and i will get to merging this and releasing it sometime tomorrow!

Commit:
testing for the SprinklesContentObserver

zsiegel commented:
Im not really sure of a good way to test the `register()` and `unregister()` methods.

Unfortunately the the ContentResolver has no API for getting the currently registered observers.

emilsjolander commented:
Yeah i know, i don't have a great idea. One way would be to check that the ContentResolver is notified when saving a model.

zsiegel commented:
I think I have an idea using mocks. Let me test a few things out.

Zachary Siegel
zsiegel87@gmail.com
On Feb 26, 2014 11:59 AM, "Emil Sjölander" notifications@github.com wrote:

> Yeah i know, i don't have a great idea. One way would be to check that the
> ContentResolver is notified when saving a model.
> 
> ## 
> 
> Reply to this email directly or view it on GitHubhttps://github.com/emilsjolander/sprinkles/pull/54#issuecomment-36155425
> .

zsiegel commented:
So my thought was to mock the `ContentObserver` object then do a `Model.save()` and verify that the `onChange()` method was called.

Without having a `SyncAdapter`, `ContentProvider`, and `AccountAuthenticator` wired into the manifest that Robolectric reads during the tests the `SprinklesContentObserver` will never get its `onChange()` methods called.

Therefore making even testing the `register()` and `unregister()` methods essentially moot.

Any other ideas?

emilsjolander commented:
Nope, i don't have any great ideas. Could we maybe just Mock a `ContentResolver` and assert that calling `register()` and `unregister()` on the `SprinklesContentObserver` actually delegates the method to the `ContentResolver`? I don't know how much this gives us though :/

zsiegel commented:
All of my attempts at trying to mock/spy the Sprinkles object with Mockito fail because of the fact that the `init` method is static and Mockito does not play nice with these kind of objects.

I think we are just going to have to live with this unless we want to investigate using something like `PowerMock` which can allegedly handle situations like this.

emilsjolander commented:
Yeah i'm ok with skipping the tests on this if we open a separate issue for it so that we have it documented and on the todo list :+1: I can create that issue after i have merge these changes.

emilsjolander commented on library/src/main/java/se/emilsjolander/sprinkles/Sprinkles.java:
```diff
@@ -8,21 +10,13 @@
-	List<Migration> mMigrations = new ArrayList<Migration>();
+
+    Context mContext;
+    List<Migration> mMigrations = new ArrayList<Migration>();
```
Could you fix the indentation here?

emilsjolander commented on sample/AndroidManifest.xml:
```diff
@@ -8,12 +8,24 @@
         android:minSdkVersion="14"
         android:targetSdkVersion="19" />
 
+    <!--Permissions used by the SyncAdapter-->
```
these permissions are no longer needed

emilsjolander commented:
I'm thinking that people may want to attach other content observers that do not have to do with a sync adapter. So how about we implement it as such that `SprinklesContentObserver` is just a wrapper around a `ContentObserver` supplied by the user? if you make this change don't forget to also update the readme :)

Good job otherwise, i like it :+1:

Commit:
refactoring SprinklesContentObserver to be a wrapper of ContentObserver

Commit:
missed a private modifier on the sync() method

zsiegel commented:
Ok so I updated the SprinklesContentObserver so that it could also take a ContentObserver supplied by the user.

The SprinklesContentObserver then just forwards those method calls to the user supplied ContentObserver. I also added some testing using Mockito to verify that those calls are being forwarded to the supplied observer.

updated the README and fixed the formatting as well. Intellij must have done some funky things on a previous check in.

thanks.

emilsjolander commented:
Do you have any good reason to a constructor specific for notifying a `SyncAdapter`? I would think that it's better not to make those kind of assumptions in this library and just let the user create their own `ContentObserver` which notifies the correct `SyncAdapter`.

zsiegel commented:
The user needs a handle to the content resolver used by Sprinkles though right? Doesn't that mean we would need to expose that somehow?

Im thinking of this looking at this line in the `SprinklesContentObserver`

``` java
Sprinkles.sInstance.mContext.getContentResolver().registerContentObserver(Utils.getNotificationUri(clazz), notifyDescendants, this);
```

emilsjolander commented:
Why would they? They would just wrap the sync functionality in a `ContentObserver` like this 

``` java
class AccountSyncObserver extends ContentObserver {

    Account mAccount;
    String mAuthority;

    AccountSyncObserver(Account account, String authority) {
        this.mAccount = account;
        this.mAuthority = authority;
    }

    @Override
    public void onChange(boolean selfChange) {
        super.onChange(selfChange);
        sync();
    }

    private void sync() {
        Bundle extras = new Bundle();
        extras.putBoolean(ContentResolver.SYNC_EXTRAS_UPLOAD, true);
        ContentResolver.requestSync(mAccount, mAuthority, extras);
    }

}
```

And than just wrap is in a `SprinklesContentObserver` like this

``` java
SprinklesContentObserver observer = new SprinklesContentObserver(new AccountSyncObserver(acc, auth))
```

Or am i missing something?

zsiegel commented:
Ahh ok now I see what you are getting at. This makes sense to me. Ill update the code and wording of the README to reflect this.

Essentially we are providing a new hook into the database. Which if the user chooses allows them the ability to launch any code (such as a SyncAdapter)

Commit:
refactoring the SprinklesContentObserver

Commit:
updating README to reflect the new behavior of the SprinklesContentObserver

Commit:
remove errant constant

emilsjolander commented:
I fixed a test that was failing and refactor some small things but nothing much. Good work, it took a while but we got a great API out of it. Will push to maven central now :+1:
